File: src/api/claude.ts
/**
 * Claude CLI API with SSE streaming - replaces Tauri claude-cli commands.
 */

import { getAuthToken } from './client';

interface ClaudeMessage {
  role: string;
  content: string;
}

interface DetectResult {
  installed: boolean;
  version?: string;
  path?: string;
  error?: string;
}

export async function detectClaudeCli(): Promise<DetectResult> {
  const token = getAuthToken();
  const response = await fetch('/api/claude/detect', {
    headers: { Authorization: `Bearer ${token}` },
  });
  return response.json();
}

/**
 * Spawn Claude CLI and stream output via SSE.
 * Returns an EventSource-like interface for consuming the stream.
 */
export function streamClaude(
  streamId: string,
  model: string,
  messages: ClaudeMessage[],
  onLine: (line: string) => void,
  onError: (error: string) => void,
  onComplete: (code: number | null, stderr: string) => void
): () => void {
  const token = getAuthToken();

  // Use fetch + ReadableStream for POST with SSE-like response
  fetch('/api/claude/spawn', {
    method: 'POST',
    headers: {
      Authorization: `Bearer ${token}`,
      'Content-Type': 'application/json',
    },
    body: JSON.stringify({ stream_id: streamId, model, messages }),
  })
    .then((response) => {
      if (!response.ok) {
        throw new Error(`HTTP ${response.status}`);
      }

      const reader = response.body?.getReader();
      if (!reader) {
        throw new Error('No response body');
      }

      const decoder = new TextDecoder();
      let buffer = '';

      // Read stream chunks
      const pump = (): Promise<void> =>
        reader.read().then(({ done, value }) => {
          if (done) {
            onComplete(null, '');
            return;
          }

          buffer += decoder.decode(value, { stream: true });

          // Split on newlines and process each SSE event
          const lines = buffer.split('\n');
          buffer = lines.pop() || ''; // Keep incomplete line in buffer

          for (const line of lines) {
            if (line.startsWith('data:')) {
              const data = line.slice(5).trim();
              if (data) {
                onLine(data);
              }
            }
          }

          return pump();
        });

      return pump();
    })
    .catch((err) => {
      onError(err.message);
    });

  // Return a cancel function (currently not implemented - would need server-side kill)
  return () => {
    // Could call /api/claude/kill here
  };
}

export async function killClaude(streamId: string): Promise<void> {
  const token = getAuthToken();
  await fetch('/api/claude/kill', {
    method: 'POST',
    headers: {
      Authorization: `Bearer ${token}`,
      'Content-Type': 'application/json',
    },
    body: JSON.stringify({ stream_id: streamId }),
  });
}

File: src-server/src/routes/claude.rs
use axum::{
    extract::State,
    response::sse::{Event, Sse},
    Json,
};
use futures::stream::Stream as StreamTrait;
use futures::StreamExt;
use serde::{Deserialize, Serialize};

use crate::AppState;
use crate::services::claude::{self, ClaudeMessage};

#[derive(Serialize)]
pub struct DetectResult {
    pub installed: bool,
    pub version: Option<String>,
    pub path: Option<String>,
    pub error: Option<String>,
}

/// Detect Claude CLI installation
pub async fn detect() -> Result<Json<DetectResult>, String> {
    let result = claude::detect().await;
    Ok(Json(DetectResult {
        installed: result.installed,
        version: result.version,
        path: result.path,
        error: result.error,
    }))
}

#[derive(Deserialize)]
pub struct SpawnBody {
    stream_id: String,
    model: String,
    messages: Vec<ClaudeMessageBody>,
}

#[derive(Deserialize)]
pub struct ClaudeMessageBody {
    role: String,
    content: String,
}

/// Spawn Claude CLI and stream output via SSE
pub async fn spawn(
    State(state): State<AppState>,
    Json(body): Json<SpawnBody>,
) -> Result<Sse<impl StreamTrait<Item = Result<Event, std::convert::Infallible>>>, String> {
    // Convert messages
    let messages: Vec<ClaudeMessage> = body.messages.into_iter().map(|m| ClaudeMessage {
        role: m.role,
        content: m.content,
    }).collect();

    let children = std::sync::Arc::clone(&state.claude_processes.children);

    let output_stream = claude::spawn(&body.model, &messages, children, body.stream_id).await?;

    // Convert string stream to SSE events
    let sse_stream = output_stream.map(|line| {
        Ok(match line {
            Ok(l) => Event::default().data(l),
            Err(_) => Event::default().data("{\"type\":\"error\"}"),
        })
    });

    Ok(Sse::new(sse_stream))
}

#[derive(Deserialize)]
pub struct KillBody {
    stream_id: String,
}

/// Kill running Claude CLI process
pub async fn kill(
    State(state): State<AppState>,
    Json(body): Json<KillBody>,
) -> Result<(), String> {
    claude::kill(std::sync::Arc::clone(&state.claude_processes.children), &body.stream_id).await?;
    Ok(())
}